File: aoc_erlang/apps/aoc_erlang/src/intcode.erl
%%% Intcode computer implementation for the 2019 puzzles.
-module(intcode).

-include_lib("eunit/include/eunit.hrl").

%% API
-export([ execute/4
        , parse/1
        ]).

%% Opcode definitions
-define(OP_ADD, 1).
-define(OP_MUL, 2).
-define(OP_INPUT, 3).
-define(OP_OUTPUT, 4).
-define(OP_JUMP_IF_TRUE, 5).
-define(OP_JUMP_IF_FALSE, 6).
-define(OP_LESS_THAN, 7).
-define(OP_EQUALS, 8).
-define(OP_ADJ_RELBASE, 9).
-define(OP_END, 99).

%% Addressing modes
-define(MODE_POS, 0).
-define(MODE_IMM, 1).
-define(MODE_REL, 2).

-type intcode_state() :: term().
-type intcode_program() :: map().

%% Input fun receives the state, and returns a new state + an input
%% value.
-type intcode_input() :: fun((State :: intcode_state())
                             -> {NewState :: intcode_state(),
                                 Input :: integer()}).

%% Output fun receives the output value and the state, and returns a
%% new state.
-type intcode_output() :: fun((Output :: integer(),
                               State :: integer())
                              -> NewState :: intcode_state()).

%% Execute an IntCode program. Input and output are communicated
%% through funs which is passed a state which can be modified and
%% returned by the input/output funs.
%%
%% @param   Prog      The IntCode program to execute.
%% @param   Input     Fun to provide input values
%% @param   Output    Fun to receive output values
%% @param   State     Initial state
%% @returns EndState  State when program exits.
-spec execute(Prog :: intcode_program(),
              Input :: intcode_input(),
              Output :: intcode_output(),
              InitState :: term()
             ) ->
                 intcode_state().
execute(Prog, Input, Output, State) ->
  execute(Prog, 0, 0, Input, Output, State).

execute(Prog, PC, RelBase, In, Out, State) ->

  %% Read
  R = fun(K) -> maps:get(K, Prog, 0) end,

  %% Write (POS and REL)
  W = fun(K, ?MODE_POS, V) -> maps:put(K, V, Prog);
         (K, ?MODE_REL, V) -> maps:put(K + RelBase, V, Prog)
      end,

  %% Read (POS, REL, and IMM)
  RM = fun(K, ?MODE_POS) -> R(K);
          (K, ?MODE_REL) -> R(K + RelBase);
          (K, ?MODE_IMM) -> K
       end,

  %% Decode the opcode into instruction and addressing mode
  Op0 = R(PC) rem 100,
  M1  = (R(PC) div 100)   rem 10,
  M2  = (R(PC) div 1000)  rem 10,
  M3  = (R(PC) div 10000) rem 10,

  %% Operands
  Op1 = R(PC + 1),
  Op2 = R(PC + 2),
  Op3 = R(PC + 3),

  case Op0 of
    ?OP_ADD ->
      execute(W(Op3, M3, RM(Op1, M1) + RM(Op2, M2)), PC + 4, RelBase, In, Out, State);
    ?OP_MUL ->
      execute(W(Op3, M3, RM(Op1, M1) * RM(Op2, M2)), PC + 4, RelBase, In, Out, State);
    ?OP_JUMP_IF_TRUE ->
      case RM(Op1, M1) of
        0 -> execute(Prog, PC + 3, RelBase, In, Out, State);
        _ -> execute(Prog, RM(Op2, M2), RelBase, In, Out, State)
      end;
    ?OP_JUMP_IF_FALSE ->
      case RM(Op1, M1) of
        0 -> execute(Prog, RM(Op2, M2), RelBase, In, Out, State);
        _ -> execute(Prog, PC + 3, RelBase, In, Out, State)
      end;
    ?OP_LESS_THAN ->
      case RM(Op1, M1) < RM(Op2, M2) of
        true  -> execute(W(Op3, M3, 1), PC + 4, RelBase, In, Out, State);
        false -> execute(W(Op3, M3, 0), PC + 4, RelBase, In, Out, State)
      end;
    ?OP_EQUALS ->
      case RM(Op1, M1) == RM(Op2, M2) of
        true  -> execute(W(Op3, M3, 1), PC + 4, RelBase, In, Out, State);
        false -> execute(W(Op3, M3, 0), PC + 4, RelBase, In, Out, State)
      end;
    ?OP_ADJ_RELBASE ->
      execute(Prog, PC + 2, RelBase + RM(Op1, M1), In, Out, State);
    ?OP_INPUT ->
      {State0, Input} = In(State),
      ?assert(is_map(State0)),
      execute(W(Op1, M1, Input), PC + 2, RelBase, In, Out, State0);
    ?OP_OUTPUT ->
      State0 = Out(RM(Op1, M1), State),
      ?assert(is_map(State0)),
      execute(Prog, PC + 2, RelBase, In, Out, State0);
    ?OP_END ->
      State
  end.

%% Parse an intcode program and return it as a map from memory
%% locations (integers) to values (integers).
-spec parse(Prog :: binary()
                  | string()) ->
               map().
parse(Binary) when is_binary(Binary) ->
  parse(binary_to_list(Binary));
parse(String) when is_list(String) ->
  Ints = string:tokens(String, ","),
  maps:from_list(
    lists:zip(lists:seq(0, length(Ints) - 1),
              lists:map(fun list_to_integer/1,
                        lists:map(fun string:trim/1, Ints)))).

File: aoc_java/src/common/IntCode.java
package common;

import static org.junit.Assert.fail;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Implementation of the Intcode machine using 64-bit ints. See e.g.
 * https://adventofcode.com/2019/day/9.
 * 
 * @author jesperes
 */
public class IntCode {

    /*
     * Opcodes
     */
    final int OP_ADD = 1;
    final int OP_MUL = 2;
    final int OP_INPUT = 3;
    final int OP_OUTPUT = 4;
    final int OP_JUMP_IF_TRUE = 5;
    final int OP_JUMP_IF_FALSE = 6;
    final int OP_LESS_THAN = 7;
    final int OP_EQUALS = 8;
    final int OP_ADJ_RELBASE = 9;
    final int OP_END = 99;

    /*
     * Operand modes
     */
    final int MODE_POS = 0;
    final int MODE_IMM = 1;
    final int MODE_REL = 2;

    /**
     * The program. We represent it as a map since the intcode spec says that
     * memory is "infinite", and any reads from uninitialized memory should
     * return 0.
     */
    private Map<Long, Long> prog;

    /** Program counter */
    private long pc = 0L;

    /** Base address for relative addressing */
    private long relbase = 0L;

    private List<Long> inputs = new ArrayList<>();
    private List<Long> outputs = new ArrayList<>();

    public static Map<Long, Long> parse(String prog) {
        Map<Long, Long> map = new HashMap<>();
        String[] strings = prog.split(",");
        for (int i = 0; i < strings.length; i++) {
            map.put((long) i, Long.valueOf(strings[i]));
        }
        return map;
    }

    public List<Long> getOutputs() {
        return outputs;
    }

    public IntCode(Map<Long, Long> prog, long... inputs) {
        this.prog = prog;
        for (long i : inputs) {
            this.inputs.add(i);
        }
        this.outputs = new ArrayList<>();
    }

    /**
     * Execute the given intcode program, supplying a list of inputs. Returns a
     * list containing the output values.
     * 
     * @param prog
     * @param inputs
     * @return
     * @return
     */
    public static List<Long> execute(String prog, long... inputs) {
        IntCode intcode = new IntCode(parse(prog), inputs);
        intcode.execute();
        return intcode.getOutputs();
    }

    private Long read(long op, int mode) {
        switch (mode) {
        case MODE_POS:
            return prog.getOrDefault(op, 0L);
        case MODE_IMM:
            return op;
        case MODE_REL:
            return prog.getOrDefault(op + relbase, 0L);
        default:
            fail();
            return 0L;
        }
    }

    private void write(Long op, int mode, Long value) {
        switch (mode) {
        case MODE_POS:
            prog.put(op, value);
            return;
        case MODE_REL:
            prog.put(op + relbase, value);
            return;
        default:
            assert false;
            return;
        }
    }

    public void execute() {

        while (true) {
            int op0 = read(pc, MODE_POS).intValue() % 100;
            int m1 = (read(pc, MODE_POS).intValue() / 100) % 10;
            int m2 = (read(pc, MODE_POS).intValue() / 1000) % 10;
            int m3 = (read(pc, MODE_POS).intValue() / 10000) % 10;

            Long op1 = read(pc + 1, MODE_POS);
            Long op2 = read(pc + 2, MODE_POS);
            Long op3 = read(pc + 3, MODE_POS);

            switch (op0) {
            case OP_ADD:
                write(op3, m3, read(op1, m1) + read(op2, m2));
                pc += 4;
                break;
            case OP_MUL:
                write(op3, m3, read(op1, m1) * read(op2, m2));
                pc += 4;
                break;
            case OP_JUMP_IF_TRUE:
                if (read(op1, m1).equals(0L))
                    pc += 3;
                else
                    pc = read(op2, m2);
                break;
            case OP_JUMP_IF_FALSE:
                if (read(op1, m1).equals(0L))
                    pc = read(op2, m2);
                else
                    pc += 3;
                break;
            case OP_LESS_THAN:
                if (read(op1, m1).compareTo(read(op2, m2)) < 0) {
                    write(op3, m3, 1L);
                    pc += 4;
                } else {
                    write(op3, m3, 0L);
                    pc += 4;
                }
                break;
            case OP_EQUALS:
                if (read(op1, m1).equals(read(op2, m2))) {
                    write(op3, m3, 1L);
                    pc += 4;
                } else {
                    write(op3, m3, 0L);
                    pc += 4;
                }
                break;
            case OP_ADJ_RELBASE:
                relbase += read(op1, m1);
                pc += 2;
                break;
            case OP_INPUT:
                write(op1, m1, input());
                pc += 2;
                break;
            case OP_OUTPUT:
                output(read(op1, m1));
                pc += 2;
                break;
            case OP_END:
                return;
            }
        }
    }

    protected void output(Long output) {
        outputs.add(output);
    }

    protected Long input() {
        return inputs.remove(0);
    }
}
